File: src/audit.ts
import { existsSync, readFileSync } from 'node:fs';
import { extractPrice, historyRows } from './analysis.ts';
import { quoteHistoryPath, readWatchlist, SENSITIVE_KEYS, type JsonObject, type JsonValue } from './storage.ts';

export type AuditFinding = {
  severity: 'warn' | 'error';
  code: string;
  message: string;
  ticker?: string;
  next_command?: string;
};

export function finding(
  severity: AuditFinding['severity'],
  code: string,
  message: string,
  options: { ticker?: string; nextCommand?: string } = {},
): AuditFinding {
  return {
    severity,
    code,
    message,
    ...(options.ticker ? { ticker: options.ticker.toUpperCase() } : {}),
    ...(options.nextCommand ? { next_command: options.nextCommand } : {}),
  };
}

function containsSensitiveKey(value: JsonValue | undefined): boolean {
  if (Array.isArray(value)) return value.some((item) => containsSensitiveKey(item));
  if (value && typeof value === 'object') {
    return Object.entries(value).some(([key, item]) => {
      const normalized = key.toLowerCase().replaceAll('-', '_');
      return SENSITIVE_KEYS.has(normalized) || containsSensitiveKey(item);
    });
  }
  return false;
}

function readJsonlForAudit(path: string, ticker: string): { records: JsonObject[]; findings: AuditFinding[] } {
  if (!existsSync(path)) return { records: [], findings: [] };
  const records: JsonObject[] = [];
  const findings: AuditFinding[] = [];
  for (const [index, line] of readFileSync(path, 'utf8').split(/\r?\n/).entries()) {
    if (!line.trim()) continue;
    try {
      const payload = JSON.parse(line) as unknown;
      if (payload && typeof payload === 'object' && !Array.isArray(payload)) records.push(payload as JsonObject);
      else findings.push(finding('error', 'malformed_record', `${path} line ${index + 1} is not an object`, { ticker, nextCommand: `history ${ticker}` }));
    } catch (error) {
      const message = error instanceof SyntaxError ? error.message : String(error);
      findings.push(finding('error', 'malformed_record', `${path} line ${index + 1} is not valid JSON: ${message}`, { ticker, nextCommand: `history ${ticker}` }));
    }
  }
  return { records, findings };
}

function validTimestamp(value: JsonValue | undefined): boolean {
  if (typeof value !== 'string' || !value.trim()) return false;
  return !Number.isNaN(Date.parse(value.replace(/Z$/, '+00:00')));
}

export function auditWatchlist(base = 'data'): AuditFinding[] {
  return readWatchlist(base).length ? [] : [
    finding('warn', 'empty_watchlist', 'watchlist is empty; add at least one ticker before running quote/history workflows', { nextCommand: '/watchlist add AAPL' }),
  ];
}

export function auditQuotes(base = 'data', ticker?: string): AuditFinding[] {
  const tickers = ticker ? [ticker.toUpperCase()] : readWatchlist(base);
  const findings: AuditFinding[] = [];
  for (const symbol of tickers) {
    const { records, findings: readFindings } = readJsonlForAudit(quoteHistoryPath(symbol, base), symbol);
    findings.push(...readFindings);
    if (!records.length) {
      if (!readFindings.length) findings.push(finding('warn', 'missing_quote_history', `no quote history found for ${symbol}`, { ticker: symbol, nextCommand: `quote ${symbol}` }));
      continue;
    }
    const seen = new Set<string>();
    records.forEach((record, index) => {
      const fetchedAt = record.fetched_at;
      if (!fetchedAt) findings.push(finding('error', 'missing_fetched_at', `record ${index} has no fetched_at timestamp`, { ticker: symbol, nextCommand: `history ${symbol}` }));
      else if (!validTimestamp(fetchedAt)) findings.push(finding('error', 'invalid_timestamp', `record ${index} has invalid fetched_at timestamp: ${String(fetchedAt)}`, { ticker: symbol, nextCommand: `history ${symbol}` }));
      else if (seen.has(String(fetchedAt))) findings.push(finding('warn', 'duplicate_timestamp', `duplicate fetched_at timestamp: ${String(fetchedAt)}`, { ticker: symbol, nextCommand: `history ${symbol}` }));
      else seen.add(String(fetchedAt));
      if (extractPrice(record) === null) findings.push(finding('error', 'missing_price', `record ${index} has no extractable price`, { ticker: symbol, nextCommand: `quote ${symbol}` }));
      if (containsSensitiveKey(record)) findings.push(finding('error', 'sensitive_key', `record ${index} contains a sensitive-looking key`, { ticker: symbol }));
    });
    const rows = historyRows(records).filter((row) => row.price !== null);
    for (let i = 1; i < rows.length; i += 1) {
      const previous = rows[i - 1]!.price!;
      const current = rows[i]!.price!;
      if (previous === 0) continue;
      const change = current / previous - 1;
      if (Math.abs(change) > 0.30) findings.push(finding('warn', 'large_price_jump', `adjacent quote price changed by ${(change * 100).toFixed(1)}%; verify source payload`, { ticker: symbol, nextCommand: `history ${symbol}` }));
    }
  }
  return findings;
}

export function auditAll(base = 'data', ticker?: string): AuditFinding[] {
  return [...(ticker ? [] : auditWatchlist(base)), ...auditQuotes(base, ticker)];
}

File: quantops_cli/audit.py
from __future__ import annotations

import json
from datetime import datetime
from pathlib import Path
from typing import Any

from .analysis import extract_price, history_rows
from .storage import SENSITIVE_KEYS, quote_history_path, read_watchlist

SENSITIVE_TERMS = SENSITIVE_KEYS


def finding(severity: str, code: str, message: str, *, ticker: str | None = None, next_command: str | None = None) -> dict[str, Any]:
    item: dict[str, Any] = {"severity": severity, "code": code, "message": message}
    if ticker:
        item["ticker"] = ticker.upper()
    if next_command:
        item["next_command"] = next_command
    return item


def _contains_sensitive_key(value: Any) -> bool:
    if isinstance(value, dict):
        for key, item in value.items():
            normalized = str(key).lower().replace("-", "_")
            if normalized in SENSITIVE_TERMS:
                return True
            if _contains_sensitive_key(item):
                return True
    if isinstance(value, list):
        return any(_contains_sensitive_key(item) for item in value)
    return False


def _read_jsonl_for_audit(path: Path, ticker: str) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    if not path.exists():
        return [], []
    records: list[dict[str, Any]] = []
    findings: list[dict[str, Any]] = []
    for line_number, line in enumerate(path.read_text(encoding="utf-8").splitlines(), start=1):
        if not line.strip():
            continue
        try:
            payload = json.loads(line)
        except json.JSONDecodeError as exc:
            findings.append(
                finding(
                    "error",
                    "malformed_record",
                    f"{path} line {line_number} is not valid JSON: {exc.msg}",
                    ticker=ticker,
                    next_command=f"history {ticker}",
                )
            )
            continue
        if not isinstance(payload, dict):
            findings.append(
                finding(
                    "error",
                    "malformed_record",
                    f"{path} line {line_number} is a {type(payload).__name__}, expected object",
                    ticker=ticker,
                    next_command=f"history {ticker}",
                )
            )
            continue
        records.append(payload)
    return records, findings


def _valid_timestamp(value: Any) -> bool:
    if not isinstance(value, str) or not value.strip():
        return False
    try:
        datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return False
    return True


def audit_watchlist(base: str | Path | None = "data") -> list[dict[str, Any]]:
    watchlist = read_watchlist(base)
    if watchlist:
        return []
    return [
        finding(
            "warn",
            "empty_watchlist",
            "watchlist is empty; add at least one ticker before running quote/history workflows",
            next_command="/watchlist add AAPL",
        )
    ]


def audit_quotes(base: str | Path | None = "data", ticker: str | None = None) -> list[dict[str, Any]]:
    watchlist = read_watchlist(base)
    tickers = [ticker.upper()] if ticker else watchlist
    findings: list[dict[str, Any]] = []
    for symbol in tickers:
        path = quote_history_path(symbol, base)
        records, read_findings = _read_jsonl_for_audit(path, symbol)
        findings.extend(read_findings)
        if not records:
            if read_findings:
                continue
            findings.append(finding("warn", "missing_quote_history", f"no quote history found for {symbol}", ticker=symbol, next_command=f"quote {symbol}"))
            continue
        seen_timestamps: set[str] = set()
        for index, record in enumerate(records):
            fetched_at = record.get("fetched_at")
            if not fetched_at:
                findings.append(finding("error", "missing_fetched_at", f"record {index} has no fetched_at timestamp", ticker=symbol, next_command=f"history {symbol}"))
            elif not _valid_timestamp(fetched_at):
                findings.append(finding("error", "invalid_timestamp", f"record {index} has invalid fetched_at timestamp: {fetched_at}", ticker=symbol, next_command=f"history {symbol}"))
            elif fetched_at in seen_timestamps:
                findings.append(finding("warn", "duplicate_timestamp", f"duplicate fetched_at timestamp: {fetched_at}", ticker=symbol, next_command=f"history {symbol}"))
            else:
                seen_timestamps.add(str(fetched_at))
            if extract_price(record.get("payload")) is None:
                findings.append(finding("error", "missing_price", f"record {index} has no extractable price", ticker=symbol, next_command=f"quote {symbol}"))
            if _contains_sensitive_key(record):
                findings.append(finding("error", "sensitive_key", f"record {index} contains a sensitive-looking key", ticker=symbol))
        rows = [row for row in history_rows(records) if row.get("price") is not None]
        for previous, current in zip(rows, rows[1:]):
            previous_price = previous.get("price")
            current_price = current.get("price")
            if previous_price in (None, 0) or current_price is None:
                continue
            change = float(current_price) / float(previous_price) - 1.0
            if abs(change) > 0.30:
                findings.append(
                    finding(
                        "warn",
                        "large_price_jump",
                        f"adjacent quote price changed by {change:.1%}; verify source payload",
                        ticker=symbol,
                        next_command=f"history {symbol}",
                    )
                )
    return findings


def audit_all(base: str | Path | None = "data", ticker: str | None = None) -> list[dict[str, Any]]:
    findings = []
    if ticker is None:
        findings.extend(audit_watchlist(base))
    findings.extend(audit_quotes(base, ticker))
    return findings
